frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+	},
+	ensure,
+	scale_info::TypeInfo,
+	traits::{
+		ChangeMembers, Currency, Get, InitializeMembers, IsSubType, OnUnbalanced,
+		ReservableCurrency,
+	},
+	weights::Weight,
+};
+use pallet_identity::IdentityField;
+
+pub use pallet::*;
+pub use types::*;
+pub use weights::*;
+
+/// Simple index type for proposal counting.
+pub type ProposalIndex = u32;
+
+type UrlOf<T, I> = BoundedVec<u8, <T as pallet::Config<I>>::MaxWebsiteUrlLength>;
+
+type BalanceOf<T, I> =
+	<<T as Config<I>>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+type NegativeImbalanceOf<T, I> = <<T as Config<I>>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+pub trait IdentityVerifier<AccountId> {
+	fn has_identity(who: &AccountId, fields: u64) -> bool;
+
+	fn has_good_judgement(who: &AccountId) -> bool;
+
+	fn super_account_id(who: &AccountId) -> Option<AccountId>;
+}
+
+impl<AccountId> IdentityVerifier<AccountId> for () {
+	fn has_identity(_who: &AccountId, _fields: u64) -> bool {
+		true
+	}
+
+	fn has_good_judgement(_who: &AccountId) -> bool {
+		true
+	}
+
+	fn super_account_id(_who: &AccountId) -> Option<AccountId> {
+		None
+	}
+}
+
+pub trait ProposalProvider<AccountId, Hash, Proposal> {
+	fn propose_proposal(
+		who: AccountId,
+		threshold: u32,
+		proposal: Box<Proposal>,
+		length_bound: u32,
+	) -> Result<(u32, u32), DispatchError>;
+
+	fn vote_proposal(
+		who: AccountId,
+		proposal: Hash,
+		index: ProposalIndex,
+		approve: bool,
+	) -> Result<bool, DispatchError>;
+
+	fn veto_proposal(proposal_hash: Hash) -> u32;
+
+	fn close_proposal(
+		proposal_hash: Hash,
+		index: ProposalIndex,
+		proposal_weight_bound: Weight,
+		length_bound: u32,
+	) -> DispatchResultWithPostInfo;
+
+	fn proposal_of(proposal_hash: Hash) -> Option<Proposal>;
+}
+
+/// The various roles that a member can hold.
+#[derive(Copy, Clone, PartialEq, Eq, RuntimeDebug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub enum MemberRole {
+	Founder,
+	Fellow,
+	Ally,
+}
+
+/// The type of item within the blacklist.
+#[derive(Clone, PartialEq, Eq, RuntimeDebug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub enum BlacklistItem<AccountId, Url> {
+	AccountId(AccountId),
+	Website(Url),
+}
+
+type BlacklistItemOf<T, I> = BlacklistItem<<T as frame_system::Config>::AccountId, UrlOf<T, I>>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The outer call dispatch type.
+		type Proposal: Parameter
+			+ Dispatchable<Origin = Self::Origin, PostInfo = PostDispatchInfo>
+			+ From<frame_system::Call<Self>>
+			+ From<Call<Self, I>>
+			+ GetDispatchInfo
+			+ IsSubType<Call<Self, I>>
```

Comment:
This basically restricts the proposal to be within the calls of this pallet, correct?

Reply:
Yes